File: tests-rs/permission-tests.rs
mod common;

use common::*;
use anchor_lang::{InstructionData, ToAccountMetas, system_program};
use private_dex::{instruction, state::LiquidityPool};
use solana_sdk::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
    signature::{Keypair, Signer},
    transaction::Transaction,
};
use litesvm_token::{CreateAssociatedTokenAccount, CreateMint, MintTo};

use crate::common::PERMISSION_PROGRAM_ID;

// ============================================================================
// Helper Functions for Permission Instructions
// ============================================================================

/// Helper: Create user permission
fn create_user_permission(
    ctx: &mut TestContext,
    payer: &Keypair,
    user_keypair: &Keypair,
    user_pda: Pubkey,
    group_id: Pubkey,
) -> Result<(Pubkey, Pubkey), Box<dyn std::error::Error>> {
    // Use utility functions for PDA derivation
    let group = get_group_pda(group_id, &PERMISSION_PROGRAM_ID);
    let permission = get_permission_pda(user_pda, group, &PERMISSION_PROGRAM_ID);
    
    let accounts = private_dex::accounts::CreateUserPermission {
        payer: payer.pubkey(),
        user: user_keypair.pubkey(),
        user_account: user_pda,
        permission,
        group,
        permission_program: PERMISSION_PROGRAM_ID,
        system_program: system_program::ID,
    };
    
    let data = instruction::CreateUserPermission { group_id };
    
    let instruction = Instruction {
        program_id: ctx.program_id,
        accounts: accounts.to_account_metas(None),
        data: data.data(),
    };
    
    let tx = Transaction::new_signed_with_payer(
        &[instruction],
        Some(&payer.pubkey()),
        &[payer],
        ctx.anchor_ctx.latest_blockhash(),
    );
    
    let signature = ctx.anchor_ctx.send_and_confirm_transaction(&tx)?;
    let tx_res = ctx.anchor_ctx.svm.get_transaction(&signature).unwrap();
    println!("tx_res: {:?}", tx_res);
    Ok((permission, group))
}

/// Helper: Create LP permission group
fn create_lp_permission_group(
    ctx: &mut TestContext,
    sender: &Keypair,
    group_id: Pubkey,
    users: Vec<Pubkey>,
) -> Result<Pubkey, Box<dyn std::error::Error>> {
    // Use utility function for PDA derivation
    let group = get_group_pda(group_id, &PERMISSION_PROGRAM_ID);
    
    let accounts = private_dex::accounts::CreateLpPermissionGroup {
        sender: sender.pubkey(),
        group,
        permission_program: PERMISSION_PROGRAM_ID,
        system_program: system_program::ID,
    };
    
    let data = instruction::CreateLpPermissionGroup { group_id };
    
    // Add remaining accounts (users to add to the group)
    let mut account_metas = accounts.to_account_metas(None);
    for user in users.iter() {
        account_metas.push(AccountMeta::new_readonly(*user, false));
    }
    
    let instruction = Instruction {
        program_id: ctx.program_id,
        accounts: account_metas,
        data: data.data(),
    };
    
    let tx = Transaction::new_signed_with_payer(
        &[instruction],
        Some(&sender.pubkey()),
        &[sender],
        ctx.anchor_ctx.latest_blockhash(),
    );
    
    ctx.anchor_ctx.send_and_confirm_transaction(&tx)?;
    Ok(group)
}

/// Helper: Create LP permission
fn create_lp_permission(
    ctx: &mut TestContext,
    sender: &Keypair,
    lp: Pubkey,
    group: Pubkey,
) -> Result<Pubkey, Box<dyn std::error::Error>> {
    // Use utility function for PDA derivation
    let permission = get_permission_pda(lp, group, &PERMISSION_PROGRAM_ID);
    
    let accounts = private_dex::accounts::CreateLpPermission {
        sender: sender.pubkey(),
        lp,
        permission,
        group,
        permission_program: PERMISSION_PROGRAM_ID,
        system_program: system_program::ID,
    };

    let lp_account = ctx.anchor_ctx.get_account::<LiquidityPool>(&lp).unwrap();
    
    let data = instruction::CreateLpPermission { mint_a: lp_account.mint_a, mint_b: lp_account.mint_b, bump: lp_account.bump };
    
    let instruction = Instruction {
        program_id: ctx.program_id,
        accounts: accounts.to_account_metas(None),
        data: data.data(),
    };
    
    let tx = Transaction::new_signed_with_payer(
        &[instruction],
        Some(&sender.pubkey()),
        &[sender],
        ctx.anchor_ctx.latest_blockhash(),
    );
    
    ctx.anchor_ctx.send_and_confirm_transaction(&tx)?;
    Ok(permission)
}

// ============================================================================
// Test Cases
// ============================================================================

#[test]
fn test_create_user_permission() {
    let mut ctx = TestContext::new();
    
    // Initialize the config
    initialize_config(&mut ctx).unwrap();
    println!("✅ Config initialized");
    
    // Create a user
    let user = Keypair::new();
    ctx.anchor_ctx.svm.airdrop(&user.pubkey(), 10_000_000_000).unwrap();
    let user_pda = create_user(&mut ctx, &user).unwrap();
    println!("✅ User created at {}", user_pda);
    
    // Create permission for the user
    let group_id = Keypair::new().pubkey(); // Random group ID
    let (permission, _group) = create_user_permission(&mut ctx, &user, &user, user_pda, group_id).expect("Failed to create user permission");
    println!("✅ User permission created at {}", permission);
}

#[test]
fn test_create_lp_permission_group() {
    let mut ctx = TestContext::new();
    
    // Initialize the config
    initialize_config(&mut ctx).unwrap();
    println!("✅ Config initialized");
    
    // Create multiple users
    let user1 = Keypair::new();
    let user2 = Keypair::new();
    let user3 = Keypair::new();
    ctx.anchor_ctx.svm.airdrop(&user1.pubkey(), 10_000_000_000).unwrap();
    ctx.anchor_ctx.svm.airdrop(&user2.pubkey(), 10_000_000_000).unwrap();
    ctx.anchor_ctx.svm.airdrop(&user3.pubkey(), 10_000_000_000).unwrap();
    
    let user1_pda = create_user(&mut ctx, &user1).unwrap();
    let user2_pda = create_user(&mut ctx, &user2).unwrap();
    let user3_pda = create_user(&mut ctx, &user3).unwrap();
    println!("✅ Users created");
    
    // Create permission group with multiple users
    let group_id = Keypair::new().pubkey();
    let users = vec![user1.pubkey(), user2.pubkey(), user3.pubkey()];
    
    create_lp_permission_group(&mut ctx, &user1, group_id, users.clone()).expect("Failed to create LP permission group");
    println!("✅ LP permission group created");
    
    println!("✅ LP permission group created successfully");
    println!("   Group PDA: {}", group_id);
    println!("   Members: {} users", users.len());
    println!("   User 1 PDA: {}", user1_pda);
    println!("   User 2 PDA: {}", user2_pda);
    println!("   User 3 PDA: {}", user3_pda);
}

#[test]
fn test_create_lp_permission() {
    let mut ctx = TestContext::new();
    
    // Initialize the config
    initialize_config(&mut ctx).unwrap();
    println!("✅ Config initialized");
    
    // Create a user to create the pool
    let creator = Keypair::new();
    ctx.anchor_ctx.svm.airdrop(&creator.pubkey(), 10_000_000_000).unwrap();
    
    // Create two token mints
    let mint_a = CreateMint::new(&mut ctx.anchor_ctx.svm, &creator)
        .authority(&creator.pubkey())
        .decimals(DECIMALS)
        .send()
        .unwrap();
    
    let mint_b = CreateMint::new(&mut ctx.anchor_ctx.svm, &creator)
        .authority(&creator.pubkey())
        .decimals(DECIMALS)
        .send()
        .unwrap();
    
    // Create liquidity pool
    let (lp, _mint_lp) = create_lp(&mut ctx, &creator, mint_a, mint_b).unwrap();
    println!("✅ Liquidity pool created at {}", lp);
    
    // First, create a permission group
    let group_id = Keypair::new().pubkey();
    let users = vec![creator.pubkey()];
    
    create_lp_permission_group(&mut ctx, &creator, group_id, users).expect("Failed to create LP permission group");
    println!("✅ LP permission group created");

    let group = get_group_pda(group_id, &PERMISSION_PROGRAM_ID);
    // Now create permission for the LP
    let permission = create_lp_permission(&mut ctx, &creator, lp, group).expect("Failed to create LP permission");
    println!("✅ LP permission created at {}", permission);

    println!("✅ LP permission created successfully");
    println!("   LP PDA: {}", lp);
    println!("   Permission PDA: {}", permission);
    println!("   Group PDA: {}", group);
}

#[test]
fn test_full_permission_flow() {
    let mut ctx = TestContext::new();
    
    println!("=== Full Permission Flow Test ===");
    
    // 1. Initialize
    initialize_config(&mut ctx).unwrap();
    println!("✅ 1. Config initialized");
    
    // 2. Create users
    let user1 = Keypair::new();
    let user2 = Keypair::new();
    ctx.anchor_ctx.svm.airdrop(&user1.pubkey(), 10_000_000_000).unwrap();
    ctx.anchor_ctx.svm.airdrop(&user2.pubkey(), 10_000_000_000).unwrap();
    
    let user1_pda = create_user(&mut ctx, &user1).unwrap();
    let user2_pda = create_user(&mut ctx, &user2).unwrap();
    println!("✅ 2. Users created");
    
    // 3. Create user permissions
    let group_id_1 = Keypair::new().pubkey();
    let group_id_2 = Keypair::new().pubkey();
    
    let perm1_result = create_user_permission(&mut ctx, &user1, &user1, user1_pda, group_id_1);
    let perm2_result = create_user_permission(&mut ctx, &user2, &user2, user2_pda, group_id_2);
    
    if perm1_result.is_ok() && perm2_result.is_ok() {
        println!("✅ 3. User permissions created");
    } else {
        println!("⚠️  3. User permission creation failed (expected if permission program not loaded)");
    }
    
    // 4. Create tokens and LP
    let mint_a = CreateMint::new(&mut ctx.anchor_ctx.svm, &user1)
        .authority(&user1.pubkey())
        .decimals(DECIMALS)
        .send()
        .unwrap();
    
    let mint_b = CreateMint::new(&mut ctx.anchor_ctx.svm, &user1)
        .authority(&user1.pubkey())
        .decimals(DECIMALS)
        .send()
        .unwrap();
    
    let (lp, _) = create_lp(&mut ctx, &user1, mint_a, mint_b).unwrap();
    println!("✅ 4. Tokens and LP created");
    
    // 5. Create LP permission group with both users
    let lp_group_id = Keypair::new().pubkey();
    let users = vec![user1.pubkey(), user2.pubkey()];
    
    create_lp_permission_group(&mut ctx, &user1, lp_group_id, users).expect("Failed to create LP permission group");
    println!("✅ LP permission group created");
    
    let group = get_group_pda(lp_group_id, &PERMISSION_PROGRAM_ID);
    
    // 6. Create LP permission
    let lp_perm_result = create_lp_permission(&mut ctx, &user1, lp, group).expect("Failed to create LP permission");
    println!("✅ LP permission created");

    println!("✅ LP permission created successfully");
    println!("   Permission PDA: {}", lp_perm_result);
    println!("   LP PDA: {}", lp);
    println!("   Group PDA: {}", group);
}

#[test]
fn test_user_permission_with_deposit() {
    let mut ctx = TestContext::new();
    
    println!("=== User Permission with Deposit Test ===");
    
    // Initialize the config
    initialize_config(&mut ctx).unwrap();
    println!("✅ Config initialized");
    
    // Create a user
    let user = Keypair::new();
    ctx.anchor_ctx.svm.airdrop(&user.pubkey(), 10_000_000_000).unwrap();
    let user_pda = create_user(&mut ctx, &user).unwrap();
    println!("✅ User created");
    
    // Create a token mint and mint tokens to the user
    let mint = CreateMint::new(&mut ctx.anchor_ctx.svm, &user)
        .authority(&user.pubkey())
        .decimals(DECIMALS)
        .send()
        .unwrap();
    
    let user_ata = CreateAssociatedTokenAccount::new(&mut ctx.anchor_ctx.svm, &user, &mint)
        .owner(&user.pubkey())
        .send()
        .unwrap();
    
    MintTo::new(&mut ctx.anchor_ctx.svm, &user, &mint, &user_ata, 10_000_000)
        .send()
        .unwrap();
    println!("✅ Tokens minted");
    
    // Deposit tokens
    deposit(&mut ctx, &user, user_pda, mint, 5_000_000).unwrap();
    println!("✅ Tokens deposited");
    
    // Create permission for the user
    let group_id = Keypair::new().pubkey();
    let (permission, group) = create_user_permission(&mut ctx, &user, &user, user_pda, group_id).expect("Failed to create user permission");
    println!("✅ User permission created");
    
    println!("✅ User permission created successfully");
    println!("   Permission PDA: {}", permission);
    println!("   Group PDA: {}", group);
    println!("   User PDA: {}", user_pda);
}

File: tests/permission-tests.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { PrivateDex } from "../target/types/private_dex";
import {
  getGroupPda,
  getPermissionPda,
  getUserPda,
  getConfigPda,
  getLiquidityPoolPda,
  PERMISSION_PROGRAM_ID,
  DELEGATE_PROGRAM_ID,
  ER_VALIDATOR_ID,
  DECIMALS,
} from "./test-utils";
import privateDexIdl from "../target/idl/private_dex.json";
import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  createMint,
  getAssociatedTokenAddressSync,
  mintToChecked,
  TOKEN_PROGRAM_ID,
  createAssociatedTokenAccountIdempotent,
} from "@solana/spl-token";
import { assert } from "chai";
import {
  Transaction,
  SystemProgram,
  PublicKey,
  Keypair,
  LAMPORTS_PER_SOL,
} from "@solana/web3.js";
import { readFileSync } from "fs";

// Load static keys for testing
const adminSecretKey = readFileSync("tmp/admin.json");
const adminKp = Keypair.fromSecretKey(new Uint8Array(JSON.parse(adminSecretKey.toString())));

describe("Permission Tests", () => {
  const wallet = new anchor.Wallet(adminKp);

  const provider = new anchor.AnchorProvider(
    new anchor.web3.Connection("http://localhost:8899", {
      wsEndpoint: "ws://localhost:8900",
      commitment: "confirmed",
    }),
    wallet
  );
  anchor.setProvider(provider);

  const program = new Program<PrivateDex>(privateDexIdl as any, provider);
  const configPda = getConfigPda(program.programId);

  before(async () => {
    // Airdrop SOL to admin
    try {
      const sig = await provider.connection.requestAirdrop(
        adminKp.publicKey,
        2 * LAMPORTS_PER_SOL
      );
      await provider.connection.confirmTransaction(sig);
    } catch (e) {
      console.log("Airdrop failed, admin might already have funds:", e.message);
    }

    // Initialize config
    try {
      await program.methods
        .initialize()
        .accountsPartial({
          sender: adminKp.publicKey,
          config: configPda,
          delegateProgram: DELEGATE_PROGRAM_ID,
          erValidator: ER_VALIDATOR_ID,
          permissionProgram: PERMISSION_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
        })
        .rpc();
      console.log("✅ Config initialized");
    } catch (e) {
      console.log("Config may already be initialized:", e.message);
    }
  });

  it("Creates user permission", async () => {
    console.log("\n=== Test: Create User Permission ===");

    const userKp = Keypair.generate();
    
    // Airdrop to user
    const sig = await provider.connection.requestAirdrop(
      userKp.publicKey,
      LAMPORTS_PER_SOL
    );
    await provider.connection.confirmTransaction(sig);

    const userPda = getUserPda(userKp.publicKey, program.programId);

    // Create user account
    await program.methods
      .createUser()
      .accountsPartial({
        sender: userKp.publicKey,
        user: userPda,
        config: configPda,
        systemProgram: SystemProgram.programId,
      })
      .signers([userKp])
      .rpc();
    console.log("✅ User created at", userPda.toString());

    // Create permission
    const groupId = Keypair.generate().publicKey;
    const groupPda = getGroupPda(groupId, PERMISSION_PROGRAM_ID);
    const permissionPda = getPermissionPda(userPda, groupPda, PERMISSION_PROGRAM_ID);

    await program.methods
      .createUserPermission(groupId)
      .accountsPartial({
        payer: userKp.publicKey,
        user: userKp.publicKey,
        userAccount: userPda,
        permission: permissionPda,
        group: groupPda,
        permissionProgram: PERMISSION_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .signers([userKp])
      .rpc();

    console.log("✅ User permission created successfully");
    console.log("   Permission PDA:", permissionPda.toString());
    console.log("   Group PDA:", groupPda.toString());
  });

  it("Creates LP permission group", async () => {
    console.log("\n=== Test: Create LP Permission Group ===");

    const user1Kp = Keypair.generate();
    const user2Kp = Keypair.generate();
    const user3Kp = Keypair.generate();

    // Airdrop to users
    for (const kp of [user1Kp, user2Kp, user3Kp]) {
      const sig = await provider.connection.requestAirdrop(
        kp.publicKey,
        LAMPORTS_PER_SOL
      );
      await provider.connection.confirmTransaction(sig);
    }

    // Create user accounts
    for (const kp of [user1Kp, user2Kp, user3Kp]) {
      const userPda = getUserPda(kp.publicKey, program.programId);
      await program.methods
        .createUser()
        .accountsPartial({
          sender: kp.publicKey,
          user: userPda,
          config: configPda,
          systemProgram: SystemProgram.programId,
        })
        .signers([kp])
        .rpc();
    }
    console.log("✅ Users created");

    // Create permission group
    const groupId = Keypair.generate().publicKey;
    const groupPda = getGroupPda(groupId, PERMISSION_PROGRAM_ID);
    const users = [user1Kp.publicKey, user2Kp.publicKey, user3Kp.publicKey];

    await program.methods
      .createLpPermissionGroup(groupId)
      .accountsPartial({
        sender: user1Kp.publicKey,
        group: groupPda,
        permissionProgram: PERMISSION_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .remainingAccounts(
        users.map((user) => ({
          pubkey: user,
          isSigner: false,
          isWritable: false,
        }))
      )
      .signers([user1Kp])
      .rpc();

    console.log("✅ LP permission group created successfully");
    console.log("   Group PDA:", groupPda.toString());
    console.log("   Members:", users.length, "users");
  });

  it("Creates LP permission", async () => {
    console.log("\n=== Test: Create LP Permission ===");

    const creatorKp = Keypair.generate();

    // Airdrop to creator
    const sig = await provider.connection.requestAirdrop(
      creatorKp.publicKey,
      LAMPORTS_PER_SOL
    );
    await provider.connection.confirmTransaction(sig);

    // Create two token mints
    const mintA = await createMint(
      provider.connection,
      creatorKp,
      creatorKp.publicKey,
      null,
      DECIMALS
    );

    const mintB = await createMint(
      provider.connection,
      creatorKp,
      creatorKp.publicKey,
      null,
      DECIMALS
    );

    // Create liquidity pool
    const [lpPda, lpPdaBump] = getLiquidityPoolPda(mintA, mintB, program.programId);
    const lpMintPda = PublicKey.findProgramAddressSync(
      [Buffer.from("lp_mint"), lpPda.toBuffer()],
      program.programId
    )[0];

    const vaultA = getAssociatedTokenAddressSync(mintA, configPda, true);
    const vaultB = getAssociatedTokenAddressSync(mintB, configPda, true);
    const vaultLp = getAssociatedTokenAddressSync(lpMintPda, configPda, true);

    await program.methods
      .createLp()
      .accountsPartial({
        sender: creatorKp.publicKey,
        lp: lpPda,
        config: configPda,
        mintA,
        mintB,
        mintLp: lpMintPda,
        vaultA,
        vaultB,
        vaultLp,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        tokenProgram: TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .signers([creatorKp])
      .rpc();
    console.log("✅ Liquidity pool created at", lpPda.toString());

    // Create permission group
    const groupId = Keypair.generate().publicKey;
    const groupPda = getGroupPda(groupId, PERMISSION_PROGRAM_ID);

    await program.methods
      .createLpPermissionGroup(groupId)
      .accountsPartial({
        sender: creatorKp.publicKey,
        group: groupPda,
        permissionProgram: PERMISSION_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .remainingAccounts([
        {
          pubkey: creatorKp.publicKey,
          isSigner: false,
          isWritable: false,
        },
      ])
      .signers([creatorKp])
      .rpc();
    console.log("✅ LP permission group created:", groupPda.toString());

    // Create LP permission
    const permissionPda = getPermissionPda(lpPda, groupPda, PERMISSION_PROGRAM_ID);

    await program.methods
      .createLpPermission(mintA, mintB, lpPdaBump)
      .accountsPartial({
        sender: creatorKp.publicKey,
        lp: lpPda,
        permission: permissionPda,
        group: groupPda,
        permissionProgram: PERMISSION_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .signers([creatorKp])
      .rpc();

    console.log("✅ LP permission created successfully");
    console.log("   Permission PDA:", permissionPda.toString());
    console.log("   LP PDA:", lpPda.toString());
    console.log("   Group PDA:", groupPda.toString());
  });

  it("Full permission flow", async () => {
    console.log("\n=== Test: Full Permission Flow ===");

    const user1Kp = Keypair.generate();
    const user2Kp = Keypair.generate();

    // Airdrop to users
    for (const kp of [user1Kp, user2Kp]) {
      const sig = await provider.connection.requestAirdrop(
        kp.publicKey,
        LAMPORTS_PER_SOL
      );
      await provider.connection.confirmTransaction(sig);
    }

    // Create users
    const user1Pda = getUserPda(user1Kp.publicKey, program.programId);
    const user2Pda = getUserPda(user2Kp.publicKey, program.programId);

    for (const [kp, pda] of [
      [user1Kp, user1Pda],
      [user2Kp, user2Pda],
    ] as const) {
      await program.methods
        .createUser()
        .accountsPartial({
          sender: kp.publicKey,
          user: pda,
          config: configPda,
          systemProgram: SystemProgram.programId,
        })
        .signers([kp])
        .rpc();
    }
    console.log("✅ 1. Users created");

    // Create user permissions
    const groupId1 = Keypair.generate().publicKey;
    const groupId2 = Keypair.generate().publicKey;

    for (const [kp, pda, groupId] of [
      [user1Kp, user1Pda, groupId1],
      [user2Kp, user2Pda, groupId2],
    ] as const) {
      const groupPda = getGroupPda(groupId, PERMISSION_PROGRAM_ID);
      const permissionPda = getPermissionPda(pda, groupPda, PERMISSION_PROGRAM_ID);

      await program.methods
        .createUserPermission(groupId)
        .accountsPartial({
          payer: kp.publicKey,
          user: kp.publicKey,
          userAccount: pda,
          permission: permissionPda,
          group: groupPda,
          permissionProgram: PERMISSION_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
        })
        .signers([kp])
        .rpc();
    }
    console.log("✅ 2. User permissions created");

    // Create tokens and LP
    const mintA = await createMint(
      provider.connection,
      user1Kp,
      user1Kp.publicKey,
      null,
      DECIMALS
    );
    const mintB = await createMint(
      provider.connection,
      user1Kp,
      user1Kp.publicKey,
      null,
      DECIMALS
    );
    console.log("✅ 3. Tokens created");

    const [lpPda, lpPdaBump] = getLiquidityPoolPda(mintA, mintB, program.programId);
    const lpMintPda = PublicKey.findProgramAddressSync(
      [Buffer.from("lp_mint"), lpPda.toBuffer()],
      program.programId
    )[0];

    const vaultA = getAssociatedTokenAddressSync(mintA, configPda, true);
    const vaultB = getAssociatedTokenAddressSync(mintB, configPda, true);
    const vaultLp = getAssociatedTokenAddressSync(lpMintPda, configPda, true);

    await program.methods
      .createLp()
      .accountsPartial({
        sender: user1Kp.publicKey,
        lp: lpPda,
        config: configPda,
        mintA,
        mintB,
        mintLp: lpMintPda,
        vaultA,
        vaultB,
        vaultLp,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        tokenProgram: TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .signers([user1Kp])
      .rpc();
    console.log("✅ 4. LP created");

    // Create LP permission group
    const lpGroupId = Keypair.generate().publicKey;
    const lpGroupPda = getGroupPda(lpGroupId, PERMISSION_PROGRAM_ID);

    await program.methods
      .createLpPermissionGroup(lpGroupId)
      .accountsPartial({
        sender: user1Kp.publicKey,
        group: lpGroupPda,
        permissionProgram: PERMISSION_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .remainingAccounts([
        {
          pubkey: user1Kp.publicKey,
          isSigner: false,
          isWritable: false,
        },
        {
          pubkey: user2Kp.publicKey,
          isSigner: false,
          isWritable: false,
        },
      ])
      .signers([user1Kp])
      .rpc();
    console.log("✅ 5. LP permission group created");

    // Create LP permission
    const lpPermissionPda = getPermissionPda(lpPda, lpGroupPda, PERMISSION_PROGRAM_ID);

    await program.methods
      .createLpPermission(mintA, mintB, lpPdaBump)
      .accountsPartial({
        sender: user1Kp.publicKey,
        lp: lpPda,
        permission: lpPermissionPda,
        group: lpGroupPda,
        permissionProgram: PERMISSION_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .signers([user1Kp])
      .rpc();
    console.log("✅ 6. LP permission created");

    console.log("\n🎉 Full permission flow completed successfully!");
  });
});

